pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,713 @@
+		///
+		/// Initial liquidity must be > 0.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: Must be T::CreatePoolOrigin
+		/// - `assets`: Asset ids tuple
+		/// - `initial_liquidity`: Corresponding initial liquidity of `assets`
+		/// - `amplification`: Pool amplification
+		/// - `fee`: trade fee to be applied in sell/buy trades
+		///
+		/// Emits `PoolCreated` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::create_pool())]
+		#[transactional]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			assets: (T::AssetId, T::AssetId),
+			initial_liquidity: (Balance, Balance),
+			amplification: u32,
+			fee: Permill,
+		) -> DispatchResult {
+			let who = T::CreatePoolOrigin::ensure_origin(origin)?;
+
+			let pool_assets: PoolAssets<T::AssetId> = assets.into();
+
+			ensure!(pool_assets.is_valid(), Error::<T>::SameAssets);
+
+			ensure!(
+				T::AmplificationRange::get().contains(&amplification),
+				Error::<T>::InvalidAmplification
+			);
+
+			for asset in (&pool_assets).into_iter() {
+				ensure!(T::AssetRegistry::exists(asset), Error::<T>::AssetNotRegistered);
+			}
+
+			ensure!(
+				initial_liquidity.0 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
+			ensure!(
+				initial_liquidity.1 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
+
+			ensure!(
+				T::Currency::free_balance(assets.0, &who) >= initial_liquidity.0,
+				Error::<T>::BalanceTooLow,
+			);
+			ensure!(
+				T::Currency::free_balance(assets.1, &who) >= initial_liquidity.1,
+				Error::<T>::BalanceTooLow,
+			);
+
+			let share_asset_ident = T::ShareAccountId::name(&pool_assets, Some(POOL_IDENTIFIER));
+
+			let share_asset = T::AssetRegistry::get_or_create_shared_asset(
+				share_asset_ident,
+				(&pool_assets).into(),
+				Balance::zero(),
+			)?;
+
+			let pool_id = PoolId(share_asset);
+
+			let pool = Pools::<T>::try_mutate(&pool_id, |maybe_pool| -> Result<PoolInfo<T::AssetId>, DispatchError> {
+				ensure!(maybe_pool.is_none(), Error::<T>::PoolExists);
+
+				let pool = PoolInfo {
+					assets: pool_assets.clone(),
+					amplification,
+					fee,
+				};
+
+				*maybe_pool = Some(pool.clone());
+
+				Ok(pool)
+			})?;
+
+			let reserves: AssetAmounts<Balance> = initial_liquidity.into();
+
+			let share_amount = calculate_add_liquidity_shares(
+				&AssetAmounts::default(),
+				&reserves,
+				T::Precision::get(),
+				amplification.into(),
+				Balance::zero(),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			ensure!(
+				share_amount >= T::MinPoolLiquidity::get(),
+				Error::<T>::InsufficientLiquidity
+			);
+
+			let pool_account = T::ShareAccountId::from_assets(&pool.assets, Some(POOL_IDENTIFIER));
+
+			T::Currency::transfer(assets.0, &who, &pool_account, initial_liquidity.0)?;
+			T::Currency::transfer(assets.1, &who, &pool_account, initial_liquidity.1)?;
+
+			T::Currency::deposit(pool_id.0, &who, share_amount)?;
```

Comment:
Wait, does this mean that I can transfer my share tokens to other users? If so the min_balance of 0 might be a problem, no?

Reply:
ED of share asset was changed and set to min pool liquidity ( instead of 0 ).